### PR TITLE
next: authn and authz in server

### PR DESCRIPTION
**Pull request by Steven-Chan**
connect #629 
after #644

TODO:
- ~~handle master key like the skygear-server~~

### CODE REVIEW

File: pkg/core/server/server.go [hunk 1]
```diff
@@ -55,9 +64,22 @@ func (s *Server) Handle(path string, hf handler.Factory) *mux.Route {
...
 
-		h.Handle(handler.Context{
-			ResponseWriter: rw,
-			Request:        r,
-		})
+		var authInfo auth.AuthInfo
+		if s.authInfoResolverFactory != nil {
+			resolver := s.authInfoResolverFactory.NewResolver(r.Context(), configuration)
+			authInfo, _ = resolver.Resolve(r)
+		}
```
Comment (rickmak) [inline pkg/core/server/server.go]: Assume `s.authInfoResolverFactory` is required.

Make the default factory in `NewServer`.

File: pkg/core/auth/authz/policy.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+	"time"
+
+	"github.com/skygeario/skygear-server/pkg/core/auth"
+	"github.com/skygeario/skygear-server/pkg/server/skyerr"
+
+	"github.com/skygeario/skygear-server/pkg/gateway/model"
+)
+
+type PolicyProvider interface {
+	ProvideAuthzPolicy(*http.Request) Policy
```
Comment (rickmak) [inline pkg/core/auth/authz/policy.go]: `ProvideAuthzPolicy` don't need to depends on the Request.

The policy should be same for all kind of request in a specific handler?

Comment (Steven-Chan) [reply pkg/core/auth/authz/policy.go]: This interface would be useful for restful api, and I imagine chat may be one of them.

Comment (Steven-Chan) [reply pkg/core/auth/authz/policy.go]: I think I will remove passing the request as param, and maybe we have a clearer picture on how to deal with it when doing chat.

File: pkg/core/server/server.go (hunk 1 above)
Comment (Steven-Chan) [reply pkg/core/server/server.go]: But the default requires `authDependency` to work, which the server itself has no knowledge of it.

**Comment by Steven-Chan** [comment]
@rickmak updated

**Pull request merged**